lib/routes/account.js

```diff
@@ -1725,14 +1719,57 @@ module.exports = function (
               uaBrowserVersion: sessionToken.uaBrowserVersion,
               uaOS: sessionToken.uaOS,
               uaOSVersion: sessionToken.uaOSVersion,
-              uaDeviceType: sessionToken.uaDeviceType
+              uaDeviceType: sessionToken.uaDeviceType,
+              uid: sessionToken.uid
+            }
+
+            if (email) {
+              // Ensure that the verification email is sent to the correct email
+              mailerOpts.email = email
             }
-          ))
+
+            return func(sessionToken, code, mailerOpts)
+          })
           .then(() => request.emitMetricsEvent(`email.${event}.resent`))
           .then(
             () => reply({}),
             reply
           )
+
+        function setVerifyCode() {
+          return db.accountEmails(sessionToken.uid)
+            .then((emails) => {
+              sessionToken.emails = emails
+
+              if (email) {
+                // If an email address is specified in payload, this is a request to verify
+                // an additional email. This should return the corresponding email code for verification.
+                emails.some((userEmail) => {
+                  if (userEmail.email === email) {
+                    code = userEmail.emailCode
+                    return true
+                  }
+                })
+              } else if (sessionToken.tokenVerificationId) {
+                code = sessionToken.tokenVerificationId
+              } else {
+                code = sessionToken.emailCode
+              }
+            })
```

Comment:
You've done well to tease these three cases apart, but the complexity here still worries me.  The right way to fix it is probably to move token verification into its own API, rather than overloading this endpoint - but that's for a separate PR!